src/Common/OptimizedRegularExpression.cpp

```diff
@@ -14,13 +14,42 @@ namespace DB
     }
 }
 
+namespace
+{
 
-template <bool thread_safe>
-void OptimizedRegularExpressionImpl<thread_safe>::analyze(
+struct Literal
+{
+    std::string literal;
+    bool prefix; /// this literal string is the prefix of the whole string.
+    bool suffix; /// this literal string is the suffic of the whole string.
+    void clear()
+    {
+        literal.clear();
+        prefix = false;
+        suffix = false;
+    }
+};
+
+using Literals = std::vector<Literal>;
+
+size_t shortest_alter_length(const Literals & literals)
+{
+    if (literals.empty()) return 0;
+    size_t shortest = ~(0);
+    for (const auto & lit : literals)
+    {
+        if (shortest > lit.literal.size())
+            shortest = lit.literal.size();
+    }
+    return shortest;
+}
+
+const char * analyzeImpl(
     std::string_view regexp,
-    std::string & required_substring,
+    const char * pos,
+    Literal & required_substring,
     bool & is_trivial,
-    bool & required_substring_is_prefix)
+    Literals & global_alters)
```

Comment:
The reader is slightly confused what "global" means in this context, e.g. is there also a "local" alternative?

Reply:
to distinguish from "current_alters". any suggestions?